cpp/src/arrow/util/async_util.cc

```diff
@@ -240,26 +239,29 @@ class AsyncTaskSchedulerImpl : public AsyncTaskScheduler {
   Future<> OnFinished() const override { return finished_; }
 
  private:
-  void ContinueTasksUnlocked(std::unique_lock<std::mutex>&& lk) {
+  void ContinueTasksUnlocked(std::unique_lock<std::mutex>* lk) {
     while (!queue_->Empty()) {
       int next_cost = std::min(queue_->Peek().cost(), throttle_->Capacity());
       std::optional<Future<>> maybe_backoff = throttle_->TryAcquire(next_cost);
       if (maybe_backoff) {
-        lk.unlock();
+        lk->unlock();
         if (!maybe_backoff->TryAddCallback([this] {
               return [this](const Status&) {
                 std::unique_lock<std::mutex> lk2(mutex_);
-                ContinueTasksUnlocked(std::move(lk2));
+                ContinueTasksUnlocked(&lk2);
               };
             })) {
-          lk.lock();
+          lk->lock();
           continue;
         }
         return;
       } else {
         std::unique_ptr<Task> next_task = queue_->Pop();
-        SubmitTaskUnlocked(std::move(next_task), std::move(lk));
-        lk.lock();
+        if (!SubmitTaskUnlocked(std::move(next_task), lk)) {
+          // We reached a terminal condition and there is no need to further continue
+          return;
+        }
+        lk->lock();
```

Comment:
Just for my own understanding, was this the crux of it? We'd submit a task, ignore the terminal condition, continue on, and due to the throttle, would schedule an asynchronous continuation. And when that task tried to run, we'd have already destroyed the scheduler?

Reply:
Hi David, yes, I think that's another way to describe this issue, the `SubmitTaskUnlocked` will create a callback and internally that callback will mark the task as done when `IsFullyFinished` returns true, after that point the dtor gets triggered and the next time we want to use the lock the mutex is already invalid.
I added a couple details more in the description of this draft PR https://github.com/apache/arrow/pull/14298#issue-1395458270

Reply:
The problem again stems from the fact that sometimes futures are already complete and so we run callbacks immediately.  Assuming that happens the loop collapses to something like...

```
lk.lock();
while (!queue.empty()) {
  task = queue.pop();
  lk.unlock();
  task.run();
  lk.lock();
  was_last_task = this.was_last_task;
  lk.unlock();
  if (was_last_task) {
    finished.MarkFinished();
  }
  lk.lock();
}
```

We didn't return after the task finished.  It was mostly harmless since we'd then soon quickly realize the queue was empty.  However, in some cases, as soon as we marked the finished callback done it would trigger the deletion of the scheduler itself (e.g. with subschedulers we do this).  In those cases it was an error to attempt to grab the lock again since the lock referred to deleted state.